src/core/opamHash.ml

```diff
@@ -82,31 +82,93 @@ let to_json s = `String (to_string s)
 let to_path (kind,s) =
   [string_of_kind kind; String.sub s 0 2; s]
 
-let compute ?(kind=default_kind) file = match kind with
-  | `MD5 -> md5 (Digest.to_hex (Digest.file file))
-  | (`SHA256 | `SHA512) as kind ->
-    try
-      if not OpamCoreConfig.(!r.use_openssl) then raise Exit else
-      match
-        OpamSystem.read_command_output ["openssl"; string_of_kind kind; file]
-      with
-      | [l] ->
-        let len = len kind in
-        make kind (String.sub l (String.length l - len) len)
-      | _ -> failwith "openssl error"
-    with OpamSystem.Command_not_found _ | Exit ->
-      make kind (OpamSHA.hash kind file)
+let compute ?target ?(kind=default_kind) file =
+  (* TODO This process is woefully inefficient. *)
+  let primary =
+    match kind with
+    | `MD5 -> md5 (Digest.to_hex (Digest.file file))
+    | (`SHA256 | `SHA512) as kind ->
+      try
+        if not OpamCoreConfig.(!r.use_openssl) then raise Exit else
+        match
+          OpamSystem.read_command_output ["openssl"; string_of_kind kind; file]
+        with
+        | [l] ->
+          let len = len kind in
+          make kind (String.sub l (String.length l - len) len)
+        | _ -> failwith "openssl error"
+      with OpamSystem.Command_not_found _ | Exit ->
+        make kind (OpamSHA.hash kind file)
+  in
+  let probably_binary name =
+    List.mem (Filename.extension name) [".zip"; ".tar"; ".gz"; ".tgz"; ".tbz"; ".txz"; ".tlz"]
+  in
+  if OpamStd.Option.map_default ((=) primary) true target || probably_binary file then
```

Comment:
I do like the `Std.Option` combinators, but `target = None || target = Some primary` would probably be easier to read here ^^.

Reply:
Oh all right, then 🙂